server/src/main/java/org/opensearch/index/shard/SearchOperationListener.java

```diff
@@ -71,6 +71,33 @@ default void onFailedQueryPhase(SearchContext searchContext) {}
      */
     default void onQueryPhase(SearchContext searchContext, long tookInNanos) {}
 
+    /**
+     * Executed before the slice execution in
+     * {@link org.opensearch.search.internal.ContextIndexSearcher#search(List, org.apache.lucene.search.Weight, org.apache.lucene.search.Collector)}.
+     * This will be called once per segment slice in concurrent search and only once in non-concurrent search.*
+     * @param searchContext the current search context
+     */
+    default void onPreSliceExecution(SearchContext searchContext) {}
+
+    /**
+     * Executed if the slice execution in
+     * {@link org.opensearch.search.internal.ContextIndexSearcher#search(List, org.apache.lucene.search.Weight, org.apache.lucene.search.Collector)} failed.
+     * This will be called once per segment slice in concurrent search and only once in non-concurrent search.*
+     * @param searchContext the current search context
+     */
+    default void onFailedSliceExecution(SearchContext searchContext) {}
+
+    /**
+     * Executed after the slice execution in
+     * {@link org.opensearch.search.internal.ContextIndexSearcher#search(List, org.apache.lucene.search.Weight, org.apache.lucene.search.Collector)} successfully finished.
+     * This will be called once per segment slice in concurrent search and only once in non-concurrent search.
```

Comment:
nit: slices are per shard not segment. How about `This will be called once per slice in concurrent search....."` ?

Reply:
I was calling it a "segment slice" to refer to it as a "slice of segments" as "slice" is sort of generic and used in other places in OpenSearch as well (ex. https://opensearch.org/docs/latest/api-reference/document-apis/reindex/)

I do see where you're coming from though, it kind of sounds like I'm saying it's executed per segment.

Reply:
We do call out that's in the context of concurrent search here though so just slices is fine here, will make the change.